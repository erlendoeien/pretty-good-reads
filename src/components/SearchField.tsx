// eslint-disable-next-line
import React, { useContext, useEffect, useState } from 'react';
import { IconButton, makeStyles, Paper } from '@material-ui/core';
import SearchIcon from '@material-ui/icons/Search';
import { useReactiveVar } from '@apollo/client';
import { observer } from 'mobx-react-lite';
import { toJS } from 'mobx';
import { BookStoreContext } from './stores/BookStore';
import { filterVar, limitVar, offsetVar, searchValueVar, sortVar } from '../cache';
import { sleep } from '../utils/sleep';

const useStyles = makeStyles({
    root: {
        padding: '2px 4px',
        display: 'flex',
        alignItems: 'center',
        margin: '0 5% 0 5%',
        width: '100%'
    },
    input: {
        marginLeft: '8px',
        flex: 1
    },
    iconButton: {
        padding: 10
    }
});

const SearchField: React.FC = observer(() => {
    const classes = useStyles();
    const [inputValue, setInputValue] = useState('');

    const searchValue = useReactiveVar(searchValueVar);
    const limit = useReactiveVar(limitVar);
    const offset = useReactiveVar(offsetVar);
    const filter = useReactiveVar(filterVar);
    const sort = useReactiveVar(sortVar);

    const bookStore = useContext(BookStoreContext);

    // Creates dependency on order when using in it hasch
    const variables = {
        searchQuery: searchValue,
        limit,
        offset,
        filter,
        sort
    };

    // Debounce type-search-update, clean up if unmounted before debounce hit
    useEffect(() => {
        const timeoutId = setTimeout(() => searchValueVar(inputValue), 500);
        return () => clearTimeout(timeoutId);
    }, [inputValue]);

    /**
     * Fetch new books based on new search from mobx store
     */
    useEffect(() => {
        const fetchData = async () => {
            if (searchValue) {
                // Add data to state For autocomplete
                await sleep(300).then(() => toJS(bookStore.fetchBooks(variables)));
            }
        };
        fetchData();
    }, [searchValue, variables, bookStore]);

    function handleSearch(e: React.MouseEvent<HTMLButtonElement>) {
        e.preventDefault();
        searchValueVar(inputValue);
    }

    return (
        <Paper component="form" className={classes.root}>
            <input
                placeholder={searchValue}
                data-cy="input"
                style={{ width: '500px', height: '40px' }}
                className={classes.input}
                type="search"
                onChange={(event: React.ChangeEvent<HTMLInputElement>): void =>
                    setInputValue(event.target.value)
                }
            />
            <IconButton
                type="submit"
                onClick={handleSearch}
                className={classes.iconButton}
                aria-label="search"
                data-cy="submit"
            >
                <SearchIcon />
            </IconButton>
        </Paper>
    );
});

export default SearchField;
