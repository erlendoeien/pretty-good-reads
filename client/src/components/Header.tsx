import { Grid } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import MenuBookIcon from '@material-ui/icons/MenuBook';
// eslint-disable-next-line
import React from 'react';
import SearchField from './SearchField';

const useStyles = makeStyles({
    header: {
        height: '500px',
        width: '100vw',
        overflow: 'hidden',
        background:
            "-webkit-linear-gradient(rgba(89,29,169, 0.8), rgba(5,25,96, 0.8)), url('/BookShelf.jpg')",
        backgroundSize: 'cover'
    },
    backgroundImg: {
        width: '100vw',
        opacity: '0.4'
    },
    title: {
        color: '#f1d4d4',
        fontSize: '3em',
        fontFamily: 'Pacifico, cursive'
    },
    icon: {
        fontSize: 70,
        color: '#f1d4d4'
    }
});

const Header: React.FC = () => {
    const classes = useStyles();

    return (
        <Grid
            direction="column"
            justify="center"
            alignItems="center"
            container
            className={classes.header}
        >
            <Grid item>
                <MenuBookIcon className={classes.icon} />
            </Grid>
            <Grid item>
                <h1 className={classes.title}>What to read next.. ?</h1>
            </Grid>
            <Grid item>
                <SearchField />
            </Grid>
        </Grid>
    );
};

export default Header;
