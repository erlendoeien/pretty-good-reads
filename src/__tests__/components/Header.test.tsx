import React from 'react';
import Header from 'src/components/Header';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import SearchField from 'src/components/SearchField';

Enzyme.configure({ adapter: new Adapter() });

describe('testing header', () => {
    it('renders correctly', () => {
        const header = shallow(<Header />);
        expect(header).toMatchSnapshot();
    });

    it('contains h1', () => {
        const header = shallow(<Header />);
        expect(header.find('h1').text()).toEqual('What to read next.. ?');
    });

    it('contains search field', () => {
        const header = shallow(<Header />);
        expect(header.containsMatchingElement(<SearchField />)).toBeTruthy();
    });
});
