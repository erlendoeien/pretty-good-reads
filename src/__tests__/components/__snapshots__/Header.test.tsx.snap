// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing header renders correctly 1`] = `
<WithStyles(ForwardRef(Grid))
  alignItems="center"
  className="makeStyles-header-1"
  container={true}
  direction="column"
  justify="center"
>
  <WithStyles(ForwardRef(Grid))
    item={true}
  >
    <MenuBookIcon
      className="makeStyles-icon-4"
    />
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    item={true}
  >
    <h1
      className="makeStyles-title-3"
    >
      What to read next.. ?
    </h1>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    item={true}
  >
    <SearchField />
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
